lib/charms/hydra/v0/oauth.py

```diff
@@ -0,0 +1,569 @@
+
+    ret = {}
+    for k, v in data.items():
+        if isinstance(v, (list, dict)):
+            try:
+                ret[k] = json.dumps(v)
+            except json.JSONDecodeError as e:
+                raise DataValidationError(f"Failed to encode relation json: {e}")
+        else:
+            ret[k] = v
+    return ret
+
+
+def _validate_data(data: Dict, schema: Dict) -> None:
+    """Checks whether `data` matches `schema`.
+
+    Will raise DataValidationError if the data is not valid, else return None.
+    """
+    try:
+        jsonschema.validate(instance=data, schema=schema)
+    except jsonschema.ValidationError as e:
+        raise DataValidationError(data, schema) from e
+
+
+@dataclass
+class ClientConfig:
+    """Helper class containing a client's configuration."""
+
+    redirect_uri: str
+    scope: str
+    grant_types: list[str]
+    audience: list[str] = field(default_factory=lambda: [])
+    token_endpoint_auth_method: str = "client_secret_basic"
+
+    def validate(self) -> None:
+        """Validate the client configuration."""
+        # Validate redirect_uri
+        if not re.match(url_regex, self.redirect_uri):
+            raise ClientConfigError(f"Invalid URL {self.redirect_uri}")
+
+        # Validate grant_types
+        for grant_type in self.grant_types:
+            if grant_type not in ALLOWED_GRANT_TYPES:
+                raise ClientConfigError(
+                    f"Invalid grant_type {grant_type}, must be one " f"of {ALLOWED_GRANT_TYPES}"
+                )
+
+        # Validate client authentication methods
+        if self.token_endpoint_auth_method not in ALLOWED_CLIENT_AUTHN_METHODS:
+            raise ClientConfigError(
+                f"Invalid client auth method {self.token_endpoint_auth_method}, "
+                f"must be one of {ALLOWED_CLIENT_AUTHN_METHODS}"
+            )
+
+
+class ClientCredentialsChangedEvent(EventBase):
+    """Event to notify the charm that the client credentials changed."""
+
+    def __init__(self, handle: Handle, client_id: str, client_secret_id: str):
+        super().__init__(handle)
+        self.client_id = client_id
+        self.client_secret_id = client_secret_id
+
+    def snapshot(self) -> Dict:
+        """Save event."""
+        return {
+            "client_id": self.client_id,
+            "client_secret_id": self.client_secret_id,
+        }
+
+    def restore(self, snapshot: Dict) -> None:
+        """Restore event."""
+        self.client_id = snapshot["client_id"]
+        self.client_secret_id = snapshot["client_secret_id"]
+
+
+class ProviderConfigChangedEvent(EventBase):
+    """Event to notify the charm that the provider's configuration changed."""
+
+
+class OAuthRequirerEvents(ObjectEvents):
+    """Event descriptor for events raised by `OAuthRequirerEvents`."""
+
+    client_credentials_changed = EventSource(ClientCredentialsChangedEvent)
+    provider_config_changed = EventSource(ProviderConfigChangedEvent)
```

Comment:
@gruyaume 
There are two sets of information in the providers databag, the provider's endpoints and the client's credentials (if a client has been registered). For most client applications it should make no difference and they should want to be notified only when the full set of information is available (provider endpoints + client credentials).

Semantically it makes sense to have two separate events, as the provider will always first place the provider_config in the databag and then the client credentials.

Do you think it makes sense to have 2 events or should we only emit one when all the information is available?

Reply:
I'd say place yourself in the charm author point of view, if there's nothing you can do with each event individually and when receiving them you need to create logic around checking whether the *other* info is also set, I'd say have only 1 event when all the info is necessary. If there's actions you can take with each of the event individually, sure let's use the 2 events.